js/src/output.ts

```diff
@@ -189,21 +182,42 @@ Shiny.addCustomMessageHandler("shinywidgets_comm_open", (msg_txt) => {
 // Basically out version of https://github.com/jupyterlab/jupyterlab/blob/d33de15/packages/services/src/kernel/default.ts#L1200-L1215
 Shiny.addCustomMessageHandler("shinywidgets_comm_msg", (msg_txt) => {
   const msg = jsonParse(msg_txt);
-  manager.get_model(msg.content.comm_id).then(m => {
+  const id = msg.content.comm_id;
+  const model = manager.get_model(id);
+  if (!model) {
+    console.error(`Couldn't handle message for model ${id} because it doesn't exist.`);
+    return;
+  }
+  model.then(m => {
     // @ts-ignore for some reason IClassicComm doesn't have this method, but we do
     m.comm.handle_msg(msg);
   });
 });
 
-// TODO: test that this actually works
+
+// Handle the closing of a widget/comm/model
 Shiny.addCustomMessageHandler("shinywidgets_comm_close", (msg_txt) => {
   const msg = jsonParse(msg_txt);
-  manager.get_model(msg.content.comm_id).then(m => {
-    // @ts-ignore for some reason IClassicComm doesn't have this method, but we do
-    m.comm.handle_close(msg)
+  const id = msg.content.comm_id;
+  const model = manager.get_model(id);
+  if (!model) {
+    console.error(`Couldn't close model ${id} because it doesn't exist.`);
+    return;
+  }
+  model.then(m => {
+    // Closing the model removes the previous view
+    m.close();
+    // .close() isn't enough to remove manager's reference to it,
+    // and apparently the only way to remove it is through the `comm:close` event
+    // https://github.com/jupyter-widgets/ipywidgets/blob/303cae4/packages/base-manager/src/manager-base.ts#L330-L337
+    // https://github.com/jupyter-widgets/ipywidgets/blob/303cae4/packages/base/src/widget.ts#L251-L253
+    m.trigger("comm:close");
```

Comment:
To match what's in that second link
```suggestion
    m.trigger("comm:close");
    m.close();
    // .close() isn't enough to remove manager's reference to it,
    // and apparently the only way to remove it is through the `comm:close` event
    // https://github.com/jupyter-widgets/ipywidgets/blob/303cae4/packages/base-manager/src/manager-base.ts#L330-L337
    // https://github.com/jupyter-widgets/ipywidgets/blob/303cae4/packages/base/src/widget.ts#L251-L253
```

Reply:
Please add a `.catch` to the end of the `.then` that least logs errors